docs/dev-guides/java-style.md

```diff
@@ -0,0 +1,166 @@
+
+*   **java.time.* – **the JavaTime library _should_ be used for date-time manipulation. Duration is especially common. For non-core extensions, consider the [https://www.threeten.org/threeten-extra/](https://www.threeten.org/threeten-extra/) library.
+*   **Lambda expressions** – _prefer_ lambda expressions for short and simple bits of code. If a lambda gets too complex, extract it into a named method.
+*   **Method references** – _prefer_ method references where possible (e.g. `someStream().map(MyClass::getBar)`).
+*   **Streams** – _prefer_ using Streams for operations that can be framed as a simple, compact stream expression. On the other hand: streams can interfere with exception handling, and complex streams may be more difficult to reason about than equivalent non-stream code, so judgement is required.
+
+### Javadoc
+
+Javadoc comments (`/** … */`) are meant to explain the intended purpose of the element they are attached to. See the [style guide](https://google.github.io/styleguide/javaguide.html) for a canonical list of when Javadoc is required.
+
+Some guidelines for writing good Javadoc:
+
+*   Spend more time documenting classes that will be frequently used or updated. Documentation mitigates against the risk of future misinterpretation.
+*   Block tags (`@param` and `@return`) are generally considered _optional_:
+    *   Use `@param` block tags when detail is required to clearly convey the meaning of each function parameter.
+    *   Use a `@return` block tag if the main text cannot clearly express the meaning of the returned value in a natural way.
+
+### Methods
+
+Some guidelines around Java methods:
+
+*   Keep methods short and simple.
+*   Avoid unwieldy parameter lists.
+*   Beware of accepting or returning null.
+*   _Prefer_ general types for parameters and specific types for return values.
+*   Avoid output parameters.
+
+See [EJ3 #49-56](https://github.com/david-sauvage/effective-java-summary#methods) for more tips.
+
+### Null
+
+Avoid nullable references where possible! Of course, this isn’t always possible, but null checks and `@Nullable` annotations should always be viewed with healthy skepticism.
+
+#### Nullable annotations
+
+@Nullable _may_ be used to indicate parameters or return values where null values are expected. When used, this annotation _should_ be in addition to (but not instead of!) Javadoc comments indicating where and when null values are expected.
+
+@NotNull and @Nonnull can be noisy; they _should not_ be used, unless there is a particular reason that readers might otherwise expect an argument to be nullable.
+
+#### Optional
+
+Optional is... optional :)
+
+`java.util.Optional` is often a good alternative to a null reference, especially as a method return type ([ref](https://stackoverflow.com/questions/26327957/should-java-8-getters-return-optional-type/26328555#26328555)). It is not a cure-all, however, and its use is neither required nor disallowed in Terra.
+
+Guidelines:
+
```

Comment:
It sounds like you're saying that nullable return values are less dangerous than nullable references (e.g. parameters). I would've almost thought the opposite: I trust my providers when calling a function more than my callers when writing one.

Reply:
That wasn't my intent. If you can point to language that seems to be implying that, I can try to re-phrase. The only preference stated here is that Optional use in function parameters is discouraged. That doesn't imply any expectation re: allowing nulls in parameters vs. return values.

Reply:
So you have `Avoid nullable references where possible! `, but the guidance didn't seem as strong for returning null.

Reply:
Ah, good point – I rephrased this slightly.